### ISSUE TITLE
2020-06-22 - 645. Set Mismatch

### ISSUE THREAD

**Issue by lvyaojia** [lvyaojia/lc-daily#22]
The set S originally contains numbers from 1 to n. But unfortunately, due to the data error, one of the numbers in the set got duplicated to another number in the set, which results in repetition of one number and loss of another number.

Given an array nums representing the data status of this set after the error. Your task is to firstly find the number occurs twice and then find the number that is missing. Return them in the form of an array.

Example 1:
```
Input: nums = [1,2,2,4]
Output: [2,3]
```
Note:
The given array size will in the range [2, 10000].
The given array's numbers won't have any order.

**Comment by lvyaojia** [comment]
``` python

from collections import defaultdict

# 使用哈希表记录出现的次数
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        s = defaultdict(int)
        
        for i in nums:
            s[i] += 1

        missing = -1
        dup = -1

        for i in range(1, len(nums)+1):
            if not s[i]:
                missing = i
            if s[i] > 1:
                dup = i
        
        return [dup, missing]

# 使用数组代替哈希表
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        checks = [0] * len(nums)

        for i in nums:
            checks[i-1] += 1
        
        missing = dup = None
        for i in range(len(checks)):
            if checks[i] == 0:
                missing = i+1
            if checks[i] > 1:
                dup = i + 1
            if missing and dup:
                break
        return [dup, missing]


# 使用求和公式: Sum = (n * (n+1)) / 2 + repeat - loss
# 通过 sum 相减得到 repeat，也可以用 hashtable 得到 repeat
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        hashtable = set(nums)
        sum_all = sum(nums)
        sum_remove_repeat = sum(hashtable)
        l = len(nums)
        loss = l * (l+1) / 2 - sum(nums) + (sum_all - sum_remove_repeat)
        return [sum_all - sum_remove_repeat, int(loss)]

```

**Comment by lvyaojia** [comment]
优化： 
使用异或
